packages/form-js-viewer/src/core/PathRegistry.js

```diff
@@ -0,0 +1,179 @@
+import { isArray } from 'min-dash';
+import { clone } from '../util';
+
+export default class PathRegistry {
+  constructor(formFieldRegistry, formFields) {
+    this._formFieldRegistry = formFieldRegistry;
+    this._formFields = formFields;
+    this._dataPaths = [];
+  }
+
+  canClaimPath(path, closed = false) {
+
+    let node = { children: this._dataPaths };
+
+    for (const segment of path) {
+
+      node = _getNextSegment(node, segment);
+
+      // if no node at that path, we can claim it no matter what
+      if (!node) {
+        return true;
+      }
+
+      // if we reach a leaf node, definitely not claimable
+      if (node.children === null) {
+        return false;
+      }
+
+    }
+
+    // if after all segments we reach a node with children, we can claim it only openly
+    return !closed;
+  }
+
+  claimPath(path, closed = false) {
+
+    if (!this.canClaimPath(path, closed)) {
+      throw new Error(`cannot claim path '${ path.join('.') }'`);
+    }
+
+    let node = { children: this._dataPaths };
+
+    for (const segment of path) {
+
+      let child = _getNextSegment(node, segment);
+
+      if (!child) {
+        child = { segment, claimCount: 1, children: [] };
+        node.children.push(child);
+      }
+      else {
+        child.claimCount++;
+      }
+
+      node = child;
+    }
+
+    if (closed) {
+      node.children = null;
+    }
+  }
+
+  unclaimPath(path) {
+
+    // verification Pass
+    let node = { children: this._dataPaths };
+    for (const segment of path) {
+      const child = _getNextSegment(node, segment);
+      if (!child) {
+        throw new Error(`no open path found for '${ path.join('.') }'`);
+      }
+      node = child;
+    }
+
+    // mutation Pass
+    node = { children: this._dataPaths };
+    for (const segment of path) {
+      const child = _getNextSegment(node, segment);
+
+      child.claimCount--;
+
+      if (child.claimCount === 0) {
+        node.children.splice(node.children.indexOf(child), 1);
+        break; // Abort early if claimCount reaches zero
+      }
+
+      node = child;
+    }
+  }
+
+  executeRecursivelyOnFields(field, fn, context = {}) {
```

Comment:
This one reads quite complicated, can we add some comments to give more context what's happening?